implement libmissing; untangles _libs dependencies

jbrockmendel opened the pull request:
In the status quo, `algos`, `hashtable`, `parsers`, and `period` depend on `lib`.  On top of that `groupby`, `join`, and `index` depend on `algos` and `intervaltree` depends on `hashtable`.  (In fairness, some of these dependency relationships are python imports, not cimports.)

`lib` depends on `tslib` which depends on essentially all of `tslibs`.  So we've got a whole bunch of dependencies (many of which aren't declared in setup.py).

But the `lib` functions needed by `algos`, `hashtable`, and `period` are basically all a small handful of null-checking `checnull`, `isnaobj`, `isnaobj2d` which depend only on `util` and `NaT`.

So this PR takes those functions and puts them into `_libs.missing`, then updates the relevant imports.

jbrockmendel committed:
implement _libs.missing; remove all cimports of libs._lib

jbrockmendel committed:
cimport checknull instead of import

jbrockmendel committed:
fixup import

jbrockmendel committed:
flake8 fixup

jreback commented on pandas/_libs/missing.pyx:
```diff
@@ -0,0 +1,176 @@
+        return get_timedelta64_value(val) == NPY_NAT
+    elif util.is_array(val):
+        return False
+    else:
```
is there a reason for not pulling in ``util._checknull`` here as well, as seems logical? (or just future PR)

jbrockmendel replied:
You mean defining it here instead of in `util`?  Or importing it into the namespace?  I'd be +1 on the former, indifferent to the latter.

jreback replied:
I think for sure *should* define it here. but then this puts missing as a dep of things like hashing.pyx. ok with it being a dep of any of the tslibs though.

```
(pandas) bash-3.2$ find pandas -name '*.pyx' | xargs grep _checknull
pandas/_libs/hashing.pyx:from util cimport _checknull
pandas/_libs/hashing.pyx:        elif _checknull(val):
pandas/_libs/interval.pyx:        if util._checknull(interval):
pandas/_libs/lib.pyx:from util cimport is_array, _checknull, _checknan
pandas/_libs/lib.pyx:        return _checknull(val)
pandas/_libs/lib.pyx:        return _checknull(val)
pandas/_libs/lib.pyx:        result[i] = val is NaT or util._checknull_old(val)
pandas/_libs/lib.pyx:                _checknull(x) and _checknull(y)):
pandas/_libs/lib.pyx:    if _checknull(val):
pandas/_libs/lib.pyx:        if _checknull(x):
pandas/_libs/lib.pyx:            if _checknull(x):
pandas/_libs/lib.pyx:            elif _checknull(y):
pandas/_libs/src/inference.pyx:        if util._checknull(val):
pandas/_libs/src/inference.pyx:        elif util._checknull(v):
pandas/_libs/src/inference.pyx:    if util._checknull(v):
pandas/_libs/src/inference.pyx:    if util._checknull(v):
pandas/_libs/src/inference.pyx:    if util._checknull(v):
pandas/_libs/src/inference.pyx:    if util._checknull(v):
pandas/_libs/src/inference.pyx:        return util._checknull(value)
pandas/_libs/src/inference.pyx:            bint is_generic_null = util._checknull(value)
pandas/_libs/tslib.pyx:from tslibs.nattype cimport _checknull_with_nat, NPY_NAT
pandas/_libs/tslib.pyx:        if _checknull_with_nat(val):
pandas/_libs/tslib.pyx:            if _checknull_with_nat(val):
pandas/_libs/tslib.pyx:        if _checknull_with_nat(val):
pandas/_libs/tslib.pyx:            if _checknull_with_nat(val):
pandas/_libs/tslib.pyx:            if _checknull_with_nat(val):
pandas/_libs/tslib.pyx:            if _checknull_with_nat(val):
pandas/_libs/tslibs/nattype.pyx:cdef inline bint _checknull_with_nat(object val):
pandas/_libs/tslibs/strptime.pyx:from nattype cimport _checknull_with_nat, NPY_NAT
pandas/_libs/tslibs/strptime.pyx:            if _checknull_with_nat(val):
pandas/_libs/tslibs/timedeltas.pyx:from nattype cimport _checknull_with_nat, NPY_NAT
pandas/_libs/tslibs/timedeltas.pyx:    if _checknull_with_nat(ts):
pandas/_libs/tslibs/timedeltas.pyx:    if _checknull_with_nat(other):
pandas/_libs/tslibs/timedeltas.pyx:        elif _checknull_with_nat(value):
```

jbrockmendel replied:
If we're moving util._checknull anyway, I'd advocate renaming it to e.g. `check_none_or_nan`

jreback replied:
sure

jbrockmendel replied:
> I think for sure should define it here. but then this puts missing as a dep of things like hashing.pyx. ok with it being a dep of any of the tslibs though.

I'll take a look and see which util funcs can be moved without messing with dependencies.

FWIW this PR already adds missing to the 'pxdfiles` key of hashtable, which cimports missing.checknull.  Previously it was an un-declared dependency on lib.

jbrockmendel replied:
Looks like util._checknull_old can be moved to missing (is used there once, nowhere else).  Let's saving util._checknull for later, since it is used in a bunch of places.

jreback replied:
ok with leaving these for later as well

jbrockmendel replied:
Sounds good.  Just pushed commit with docstrings.

jreback commented on pandas/_libs/missing.pyx:
```diff
@@ -0,0 +1,176 @@
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def isnaobj(ndarray arr):
+    cdef:
```
ideally add some doc-strings

jreback commented on pandas/_libs/missing.pyx:
```diff
@@ -0,0 +1,176 @@
+cdef int64_t NPY_NAT = util.get_nat()
+
+
+cdef inline bint is_null_datetimelike(v):
```
prob should rename this for consistency (checknull_datetimelike), can be TODO

jreback commented on pandas/_libs/src/inference.pyx:
```diff
@@ -533,22 +533,6 @@ cpdef object infer_datetimelike_array(object arr):
     return 'mixed'
 
 
-cdef inline bint is_null_datetimelike(v):
```
is there a reason these are not in nattype?  (and maybe namespaces to missing, but actual definition in nattype)

jbrockmendel replied:
I'd be fine putting this in nattype.  The one below I considered moving but decided against because it's only used once here in inference.pyx.

jreback commented on pandas/_libs/tslib.pyx:
```diff
@@ -830,24 +829,6 @@ class Timestamp(_Timestamp):
 
 
-cdef inline bint _check_all_nulls(object val):
-    """ utility to check if a value is any type of null """
```
is this used anywhere?

jbrockmendel replied:
Yes, its used once here in _libs.missing.

jreback commented on pandas/tests/test_lib.py:
```diff
@@ -208,14 +208,14 @@ class TestNAObj(object):
 
     def _check_behavior(self, arr, expected):
```
should move these tests to test_missing

jbrockmendel replied:
Sure.  Existing test_missing in tests.dtypes or a new test_libmissing?

jreback replied:
i think tests.dtypes is ok, maybe add test_libmissing at later point (orthogonal to this)

jbrockmendel committed:
Move libmissing tests to test_missing

jbrockmendel committed:
move util._checknull_old --> missing._check_none_nan_inf_neginf

jbrockmendel committed:
add docstrings

jreback commented on pandas/io/formats/format.py:
```diff
@@ -1860,7 +1860,7 @@ def _format_strings(self):
 
         def _format(x):
-            if self.na_rep is not None and lib.checknull(x):
+            if self.na_rep is not None and libmissing.checknull(x):
```
in followup (or here), change this to

``is_scalar(x) and isna(x)``; this is reaching too much into the internals.

jreback replied:
and all below here

*maybe* worth defining a ``is_scalar_na`` function

jreback commented on pandas/core/dtypes/missing.py:
```diff
@@ -50,7 +50,7 @@ def isna(obj):
 
 def _isna_new(obj):
     if is_scalar(obj):
-        return lib.checknull(obj)
```
see this here is ok, as this is really the python api for missing values.

jreback commented on pandas/io/formats/excel.py:
```diff
@@ -381,12 +381,12 @@ def __init__(self, df, na_rep='', float_format=None, cols=None,
 
     def _format_value(self, val):
-        if lib.checknull(val):
+        if libmissing.checknull(val):
```
use is_scalar and isna

jreback commented on pandas/io/formats/excel.py:
```diff
@@ -381,12 +381,12 @@ def __init__(self, df, na_rep='', float_format=None, cols=None,
             val = self.na_rep
         elif is_float(val):
-            if lib.isposinf_scalar(val):
+            if libmissing.isposinf_scalar(val):
```
isposinf_scalar should be exposed in core/dtypes/missing.py as a python function

jreback commented on pandas/tests/dtypes/test_missing.py:
```diff
@@ -333,3 +335,52 @@ def test_na_value_for_dtype():
+    _1d_methods = ['isnaobj', 'isnaobj_old']
+    _2d_methods = ['isnaobj2d', 'isnaobj2d_old']
+
+    def _check_behavior(self, arr, expected):
```
if you can parametrize/use fixtures would be good

jbrockmendel commented on pandas/_libs/algos_rank_helper.pxi.in:
```diff
@@ -83,7 +83,7 @@ def rank_1d_{{dtype}}(object in_arr, ties_method='average', ascending=True,
 
     {{if dtype == 'object'}}
-    mask = lib.isnaobj(values)
+    mask = missing.isnaobj(values)
```
ATM these are only not `cdef`

jbrockmendel committed:
edits per reviewer request

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into libmissing

jbrockmendel committed:
fixup missing import

jreback commented:
rebase

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into libmissing

jreback commented:
pls rebase

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into libmissing

jbrockmendel committed:
remove imports no longer used after merge

jreback commented on pandas/_libs/missing.pyx:
```diff
@@ -0,0 +1,310 @@
+cimport util
+
+from tslibs.np_datetime cimport get_timedelta64_value, get_datetime64_value
+from tslibs.nattype import NaT, iNaT
```
instead of importing iNaT, just use NPY_NAT to avoid perf issues.

jreback commented on pandas/_libs/missing.pyx:
```diff
@@ -0,0 +1,310 @@
+cdef int64_t NPY_NAT = util.get_nat()
+
+
+cdef inline bint is_null_datetimelike(v):
```
type as object

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into libmissing

jbrockmendel committed:
use NPY_NAT instead of iNaT, range instead of C syntax

jbrockmendel commented:
> instead of importing iNaT, just use NPY_NAT to avoid perf issues.

> type as object

Just pushed with both these changes.  Also: 

- changed variable `v` --> `val` (there's a recent issue regarding linting too-short variable names that I _very_ much agree with)
- changed `for i from 0 <= i < n:`  --> `for i in range(n):` in a couple places

jreback commented on setup.py:
```diff
@@ -516,6 +517,10 @@ def pxd(name):
     '_libs.lib': {
         'pyxfile': '_libs/lib',
         'depends': lib_depends + tseries_depends},
```
need to add missing to lib as well

jbrockmendel replied:
Can do.  At the moment we've basically only implemented the pxfiles "rule" for tslibs.  Should we go through and fill them out for the others?

In case I haven't mentioned it recently: I'm _really_ looking forward to using cythonize and not worrying about this.

jbrockmendel committed:
add pxdfiles for _libs.lib

jreback commented:
thanks!

jreback commented:
not sure this started showing up on this PR, but pls add to the list
```
/Users/jreback/miniconda3/envs/pandas/lib/python3.6/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: "Using deprecated NumPy API, disable it by "          "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-W#warnings]
#warning "Using deprecated NumPy API, disable it by " \
 ^
pandas/_libs/src/datetime/np_datetime.c:571:12: warning: expression result unused [-Wunused-value]
  meta.num - 1;
  ~~~~~~~~ ^ ~
pandas/_libs/src/datetime/np_datetime.c:1004:15: warning: unused variable 'perday' [-Wunused-variable]
    npy_int64 perday;
              ^
3 warnings generated.
```

jbrockmendel commented:
> not sure this started showing up on this PR, but pls add to the list

Looks like these are in the new timedelta_struct functions, likely leftovers from copy/paste.  Easy to fix in-place, but I'd advocate not having them in src/ to begin with.  Put them directly in tslibs.np_datetime (or tslibs.np_timedelta).
